### PR TITLE
feat: add InterRecordBoundedCumulatedDifferenceProcessor

### PR DESCRIPTION
**Pull request by olivierabz**
Requested by customers, developed with customer but now added as a processor.
Adaptation of the cumulated difference processor to ensure the results are within expected bounds. The logic is the same as for `InterRecordBoundedRangeDifferenceProcessor`

### CODE REVIEW

File: avatars/processors/inter_record_bounded_cumulated_difference_test.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        new_difference_variable_name="value_difference",
+    )
+
+    with pytest.raises(
+        ValueError, match="Expected no missing values for target variable in source"
+    ):
+        processor.postprocess(df_with_cumulated, preprocessed_df_with_cumulated)
+
+
+def test_postprocess_raises_error_when_missing_dest_first_var(
```
Comment (mguillaudeux) [inline avatars/processors/inter_record_bounded_cumulated_difference_test.py:101]:
```suggestion
def test_postprocess_raises_error_when_missing_in_dest_first_var(
```

File: avatars/processors/inter_record_bounded_cumulated_difference_test.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        new_difference_variable_name="value_difference",
+    )
+
+    with pytest.raises(
+        ValueError, match="Expected no missing values for `new_first_variable_name`"
+    ):
+        processor.postprocess(df_with_cumulated, preprocessed_df_with_cumulated)
+
+
+def test_preprocess_raises_error_when_missing_dest_difference_var(
```
Comment (mguillaudeux) [inline avatars/processors/inter_record_bounded_cumulated_difference_test.py:119]: same

File: avatars/processors/inter_record_bounded_cumulated_difference.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                "Expected no missing values for target variable, got column with nulls instead"
+            )
+
+        working = df.copy()
+
+        # determine lb and ub
+        working["lb"] = min(working[self.target_variable])
+        working["ub"] = max(working[self.target_variable])
+
+        # store first value of target variable
```
Comment (mguillaudeux) [inline avatars/processors/inter_record_bounded_cumulated_difference.py:114]: we might lack a unit test to check if the first value per id number is working as we expect. Don't know if it's worth making a function out of it though.

File: avatars/processors/inter_record_bounded_cumulated_difference.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        working[self.new_first_variable_name] = working.groupby(self.id_variable)[
+            self.target_variable
+        ].transform("first")
+
+        # store the difference between current and previous value
+        working["previous_val"] = working.groupby(self.id_variable)[
+            self.target_variable
+        ].shift()
```
Comment (tcrasset) [inline avatars/processors/inter_record_bounded_cumulated_difference.py:122]: nice use of `shift()` and `transform()`

**Review by tcrasset — APPROVED** [review]
You are creating lots of temporary columns, which in itself is fine. Could you add a test to make sure that you are removing all the unnecessary ones once everything is done?

File: avatars/processors/inter_record_bounded_cumulated_difference.py (hunk 3 above)
Comment (olivierabz) [reply avatars/processors/inter_record_bounded_cumulated_difference.py:114]: added an assert to test that preprocess is as expected (and so contains the expected `first_value`

**Pull request merged**